crates/bevy_ecs/src/system/commands/entity_command.rs

```diff
@@ -74,96 +76,75 @@ use bevy_ptr::OwningPtr;
 ///     assert_eq!(names, HashSet::from_iter(["Entity #0", "Entity #1"]));
 /// }
 /// ```
-///
-/// # Note on Generic
-///
-/// The `Marker` generic is necessary to allow multiple blanket implementations
-/// of `EntityCommand` for closures, like so:
-/// ```ignore (This would conflict with the real implementations)
-/// impl EntityCommand for FnOnce(Entity, &mut World)
-/// impl EntityCommand<World> for FnOnce(EntityWorldMut)
-/// impl EntityCommand<Result> for FnOnce(Entity, &mut World) -> Result
-/// impl EntityCommand<(World, Result)> for FnOnce(EntityWorldMut) -> Result
-/// ```
-/// Without the generic, Rust would consider the implementations to be conflicting.
-///
-/// The type used for `Marker` has no connection to anything else in the implementation.
-pub trait EntityCommand<Marker = ()>: Send + 'static {
+pub trait EntityCommand<T = ()>: Send + 'static {
     /// Executes this command for the given [`Entity`] and
     /// returns a [`Result`] for error handling.
-    fn apply(self, entity: Entity, world: &mut World) -> Result;
+    fn apply(self, entity: EntityWorldMut) -> T;
+}
+/// Passes in a specific entity to an [`EntityCommand`], resulting in a [`Command`] that
+/// internally runs the [`EntityCommand`] on that entity.
+pub trait CommandWithEntity<T> {
```

Comment:
Comments here on why this rather convoluted design was needed would be useful. A link back to this PR thread would be sufficient even.